File: backend/controllers/student_controller.py
from datetime import datetime
from flask import request, jsonify, g
from controllers import handle_errors
import services.student_service as svc


def _parse_date(s):
    if not s:
        return None
    for fmt in ("%Y-%m-%d", "%d/%m/%Y", "%d/%m/%y"):
        try:
            return datetime.strptime(s, fmt).date()
        except ValueError:
            pass
    return None


@handle_errors
def search_students():
    q        = request.args.get("q", "").strip()
    class_id = request.args.get("class_id", type=int)
    if not q:
        return jsonify({"success": True, "data": []})
    return jsonify({"success": True, "data": svc.search_students(q, class_id)})


@handle_errors
def add_to_class(class_id):
    username = (request.json or {}).get("username", "").strip()
    if not username:
        return jsonify({"success": False, "message": "Thiếu username học sinh"}), 400
    full_name = svc.add_to_class(class_id, g.user["user_id"], username)
    return jsonify({"success": True, "message": f"Đã thêm {full_name} vào lớp"})


@handle_errors
def upload_students(class_id):
    if "file" not in request.files:
        return jsonify({"success": False, "message": "Không tìm thấy file"}), 400
    file = request.files["file"]
    if not file.filename or not file.filename.lower().endswith((".xlsx", ".xls")):
        return jsonify({"success": False, "message": "Chỉ hỗ trợ file .xlsx hoặc .xls"}), 400
    result = svc.upload_students(class_id, g.user["user_id"], file.read())
    msg = f"Thêm thành công {result['count']} học sinh."
    if result["errors"]:
        msg += f" Có {len(result['errors'])} lỗi: " + "; ".join(result["errors"][:3])
    return jsonify({"success": True, "message": msg, "count": result["count"]})


@handle_errors
def remove_from_class(class_id, student_id):
    svc.remove_from_class(class_id, g.user["user_id"], student_id)
    return jsonify({"success": True, "message": "Đã gỡ học sinh khỏi lớp"})


@handle_errors
def get_student_results(student_id):
    return jsonify({"success": True, "data": svc.get_student_results(student_id)})


@handle_errors
def get_student_progress(student_id):
    return jsonify({"success": True, "data": svc.get_student_progress(student_id)})


@handle_errors
def student_dashboard():
    uid       = g.user["user_id"]
    show_all  = request.args.get("all") == "true"
    date_from = _parse_date(request.args.get("dateFrom", ""))
    date_to   = _parse_date(request.args.get("dateTo", ""))
    if not show_all and date_from and date_to and date_from > date_to:
        return jsonify({"success": False, "message": "dateFrom phải nhỏ hơn hoặc bằng dateTo"}), 400
    return jsonify({"success": True, "data": svc.get_student_dashboard(uid, show_all, date_from, date_to)})

File: backend/src/controllers/student.controller.js
import * as svc from '../services/student.service.js';
import { success, successMsg } from '../utils/response.js';

export const downloadSampleStudents = async (req, res) => {
  const buffer = svc.generateSampleExcel();
  res.setHeader('Content-Type', 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet');
  res.setHeader('Content-Disposition', 'attachment; filename="mau_import_hocsinh.xlsx"');
  res.send(buffer);
};

export const searchStudents = async (req, res) => {
  const q = (req.query.q || '').trim();
  const classId = req.query.classId ? parseInt(req.query.classId) : null;
  if (!q) return success(res, []);
  success(res, await svc.searchStudents(q, classId));
};

export const addToClass = async (req, res) => {
  const { username } = req.body;
  const fullName = await svc.addToClass(parseInt(req.params.classId), req.user.id, username.trim());
  successMsg(res, `Đã thêm ${fullName} vào lớp`);
};

export const uploadStudents = async (req, res) => {
  if (!req.file) return res.status(400).json({ success: false, message: 'Không tìm thấy file' });
  const result = await svc.uploadStudents(parseInt(req.params.classId), req.user.id, req.file.buffer);
  let msg = `Thêm thành công ${result.count} học sinh.`;
  if (result.errors.length) msg += ` Có ${result.errors.length} lỗi: ` + result.errors.slice(0, 3).join('; ');
  successMsg(res, msg);
};

export const removeFromClass = async (req, res) => {
  await svc.removeFromClass(parseInt(req.params.classId), req.user.id, parseInt(req.params.studentId));
  successMsg(res, 'Đã gỡ học sinh khỏi lớp');
};

export const getStudentResults = async (req, res) => {
  success(res, await svc.getStudentResults(parseInt(req.params.id)));
};

export const getStudentProgress = async (req, res) => {
  success(res, await svc.getStudentProgress(parseInt(req.params.id)));
};

export const studentDashboard = async (req, res) => {
  const showAll = req.query.all === 'true';
  const dateFrom = req.query.dateFrom ? new Date(req.query.dateFrom) : null;
  const dateTo = req.query.dateTo ? new Date(req.query.dateTo) : null;
  if (!showAll && dateFrom && dateTo && dateFrom > dateTo)
    return res.status(400).json({ success: false, message: 'dateFrom phải nhỏ hơn hoặc bằng dateTo' });
  success(res, await svc.getStudentDashboard(req.user.id, showAll, dateFrom, dateTo));
};
